### ISSUE TITLE
Kubelet's logic of checking docker running in host does not work in systemd distro

### ISSUE THREAD

**Issue by andyzheng0831** [kubernetes/kubernetes#23956]
I notice that kubelet does not create "/docker-daemon" cgroup if docker is managed by systemd, because the function isProcessRunningInHost call (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/container_manager_linux.go#L448) returns false. As a result, test cases in https://github.com/kubernetes/kubernetes/blob/master/test/e2e/kubelet_perf.go will all fail. This isProcessRunningInHost function (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/container_manager_linux.go#L408) compares /proc/1/ns/mnt with /proc/<docker pid>/ns/mnt to determine if docker is running in host. But this is not true if docker is managed by systemd (such as in CoreOS). You can see that docker has different mount namespace from pid 1, if docker is managed by systemd. But docker and pid 1 process have the same pid namespace.

Why would kubelet check the mount namespace, instead of pid namespace? Could you fix this issue in kubelet container manager?

@dchen1107 @vishh 

cc/ @roberthbailey @fabioy @zmerlynn @wonderfly FYI.

**Comment by vishh** [comment]
Never ending struggle with systemd :disappointed: 

If docker is managed by systemd, there is no need for the kubelet to move it around. Kubelet will auto-discover docker daemon's cgroups.

The reason for going with mount ns is that as of now docker can be run inside a container and one can skip all namespaces, but not the mount namespace when running inside a docker container. I don't think we can use pid ns for this purpose.

**Comment by vishh** [comment]
Don't set `--runtime-cgroups` flag in kubelet (default behavior) and it should all just work. Closing this issue as working as intended. Feel free to re-open if you have any other issues.

**Comment by dchen1107** [comment]
Looks like a regression caused by https://github.com/kubernetes/kubernetes/pull/20671. But I believe this is workaround to solve this issue by using proper flag.

**Comment by dchen1107** [comment]
Ahh, @vishh Just provided you the workaround above. ;-)

**Comment by andyzheng0831** [comment]
Will skipping "--runtime-cgroups" flag have any sideeffect in kubelet functionality? And this will make our stuff pass the kubelet_perf tests?

**Comment by vishh** [comment]
kubelet perf tests are using [old stats apis](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/kubelet_stats.go#L450). They need to be ported over to use the summary API. I'd recommend filing an issue for making the kubelet_perf.go tests be more compatible with other distros/setups.
cc @yujuhong

**Comment by andyzheng0831** [comment]
I am removing the "--runtime-cgroups" kubelet flag and run e2e tests to see if any new problems happen. @vishh based your understanding, does unsetting this flag cause any problem such as affecting any functionality? Just to double confirm in addition to running e2e.

I will file a new issue about making kubelet_perf tests better

**Comment by dchen1107** [comment]
Question: Why we spot such issue so late? Shouldn't test failed since day one to run?

**Comment by andyzheng0831** [comment]
The failed three tests are 
- Kubelet [Serial] [Slow] regular resource usage tracking for 35 pods per node over 20m0s
- Kubelet [Serial] [Slow] regular resource usage tracking for 100 pods per node over 20m0s
- Kubelet [Serial] [Slow] regular resource usage tracking for 0 pods per node over 20m0s

They are only run in GKE tests. In our k8s release and slow Jenkins, these three tests are filtered. Until recently, we have GKE tests running against our stuff, as it took us quite a long time to make everything ready. As a result, we noticed the failure in recent few days.

**Comment by dchen1107** [comment]
Why those tests are only run in GKE tests? Shouldn't they are included for regular GCE test suites too? I thought you guys test the image on top of GCE too.

**Comment by andyzheng0831** [comment]
We have forked a part of GCE and GKE based Jenkins jobs. For whatever we folk, we used the same Jenkins configuration as ContainerVM. If you check https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e.yaml, you will see these three tests are only run in the gce-serial and gke-serial. More Jenkins jobs against ContainerVM also do not run these three tests. We created our gke-serial (but not gce-serial) recently, so we hit the failure a couple of days ago. I will log an issue in my team to set up gce-serial jenkins tests.

**Comment by wonderfly** [comment]
> Why those tests are only run in GKE tests? Shouldn't they are included for regular GCE test suites too? I thought you guys test the image on top of GCE too.

We didn't replicate all e2e tests, for 1) some of them were not considered _critical_ (like disruptive and feature tests), and 2) the e2e test framework, Jenkins configs, and tests themselves changed rapidly and we were often left behind.

That said, we are planning to run as many tests against Trusty as we do for ContainerVM. We have an internal bug tracking this.

**Comment by dchen1107** [comment]
But kubernetes-e2e-gce-serial including above three tests are part of critical builds: http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-serial/ 

Anyway, I just want to make sure we are running into this issue next time. Thanks!

**Comment by dchen1107** [comment]
cc/ @kubernetes/goog-testing This is an example showing how hard to follow for our partner teams / community. We moved those slow tests to serial suite, and put them into our critical builds, without blocking merges. All are good and reasonable decision here, but not very clearly documented.

**Comment by derekwaynecarr** [comment]
It's never correct to set the runtime flag on a systemd distro for the reasons discussed since everything is already in a cgroup, there is no reason for Kubelet to have to move it.  i would love to fail Kubelet startup on systemd if this flag is set.  Any objections?

**Comment by derekwaynecarr** [comment]
The only valid cgroup flag on systemd in my view is cgroup-parent.

**Comment by yujuhong** [comment]
> kubelet perf tests are using old stats apis. They need to be ported over to use the summary API. I'd recommend filing an issue for making the kubelet_perf.go tests be more compatible with other distros/setups.
> cc @yujuhong

I've uploaded a PR #24003 for that. Unfortunately, the summary API does not include RSS memory, which is required by the test. I filed an issue for that #24005, but I don't think we want to patch v1.2 with it? In that case, the memory checking part of the test will always passes in the v1.2 branch.

**Comment by yujuhong** [comment]
@dchen1107 @vishh and I spoke offline, and we decided to disable docker-daemon checks in the test for the 1.2 branch (#24013). The PR is smaller and simpler. 

#24003 will not be cherry-picked to 1.2

**Comment by andyzheng0831** [comment]
My fix #23996 was merged and #24060 was cherry picked in release-1.2, @wonderfly also added Jenkins jobs for serial tests. The fix on my and @wonderfly side is done, and there is another issue for @yujuhong to fix the test case. So, close this issue
